packages/replayio/src/utils/recordings/upload/uploadSourceMaps.ts

```diff
@@ -0,0 +1,91 @@
+import { addOriginalSource } from "../../protocol/api/addOriginalSource";
+import { addSourceMap } from "../../protocol/api/addSourceMap";
+import { checkIfResourceExists } from "../../protocol/api/checkIfResourceExists";
+import { createResource } from "../../protocol/api/createResource";
+import { getResourceToken } from "../../protocol/api/getResourceToken";
+import { debug } from "../debug";
+import { LocalRecording } from "../types";
+
+async function ensureResource(client: ProtocolClient, content: string) {
+  const { token } = await getResourceToken(client, { hash: `sha256:${hashValue(content)}` });
+  const resource = {
+    token,
+    saltedHash: `sha256:${hashValue(token + content)}`,
+  };
+  const { exists } = await checkIfResourceExists(client, {
+    resource,
+  });
+  if (exists) {
+    return resource;
+  }
+  return (await createResource(client, { content })).resource;
+}
+
+export async function uploadSourceMaps(client: ProtocolClient, recording: LocalRecording) {
+  return promiseMap(
+    recording.metadata.sourcemaps,
+    async sourceMap => {
+      debug("Uploading sourcemap %s for recording %s", sourceMap.path, recording.id);
+      let sourceMapId: string;
+      try {
+        const sourceMapContent = await fs.readFile(sourceMap.path, "utf-8");
+        const result = await addSourceMap(client, {
+          recordingId: recording.id,
+          baseURL: sourceMap.baseURL,
+          targetContentHash: sourceMap.targetContentHash,
+          targetURLHash: sourceMap.targetURLHash,
+          targetMapURLHash: sourceMap.targetMapURLHash,
+          resource: await ensureResource(client, sourceMapContent),
+        });
+        sourceMapId = result.id;
+      } catch (error) {
+        debug(
+          "Failed to upload sourcemap %s for recording %s: %o",
+          sourceMap.path,
+          recording.id,
+          error
+        );
+        return;
+      }
+
+      return promiseMap(
+        sourceMap.originalSources,
+        async source => {
+          debug(
+            "Uploading original source %s for sourcemap %s for recording %s",
+            source.path,
+            sourceMap.path,
+            recording.id
+          );
+          try {
+            const sourceContent = await fs.readFile(source.path, "utf-8");
+            await addOriginalSource(client, {
+              recordingId: recording.id,
+              parentId: sourceMapId,
+              parentOffset: source.parentOffset,
+              resource: await ensureResource(client, sourceContent),
+            });
+          } catch (error) {
+            debug(
+              "Failed to upload original source %s for sourcemap %s for recording %s: %o",
+              source.path,
+              sourceMap.path,
+              recording.id,
+              error
+            );
+          }
+        },
+        {
+          concurrency: 5,
+        }
+      );
+    },
+    {
+      concurrency: 10,
+    }
+  );
```

Comment:
I reused this structure from the old CLI but I think it can be improved and I will follow up with a different queue-based take on this.